### PR TITLE
refactor: move {`epoll`, `kqueue`} (de)init logic and wakeup pipes logic out of `CConnman` and into `EdgeTriggeredEvents` and `WakeupPipes`

### PR DESCRIPTION
**Pull request by kwvg**
## Motivation

`CConnman` is an entity that contains a lot of platform-specific implementation logic, both inherited from upstream and added upon by Dash (support for edge-triggered socket events modes like `epoll` on Linux and `kqueue` on FreeBSD/Darwin).

Bitcoin has since moved to strip down `CConnman` by moving peer-related logic to the `Peer` struct in `net_processing` (portions of which are backported in #5982 and friends, tracking efforts from https://github.com/bitcoin/bitcoin/issues/19398) and moving socket-related logic to `Sock` (portions of which are aimed to be backported in #6004, tracking efforts from https://github.com/bitcoin/bitcoin/pull/21878).

Due to the direction being taken and the difference in how edge-triggered events modes operate (utilizing interest lists and events instead of iterating over each socket) in comparison to level-triggered modes (which are inherited from upstream), it would be reasonable to therefore, isolate Dash-specific code into its own entities and minimize the information `CConnman` has about its internal workings. 

One of the visible benefits of this approach is comparing `develop` (as of this writing, d44b0d5dcb9b54821d582b267a8b92264be2da1b) and this pull request for interactions between wakeup pipes logic and {`epoll`, `kqueue`} logic.

This is what construction looks like:

https://github.com/dashpay/dash/blob/d44b0d5dcb9b54821d582b267a8b92264be2da1b/src/net.cpp#L3358-L3397

But, if we segment wakeup pipes logic (that work on any platform with POSIX APIs and excludes Windows) and {`epoll`, `kqueue`} logic (calling them `EdgeTriggeredEvents` instead), construction looks different:

https://github.com/dashpay/dash/blob/907a3515170abed4ce9018115ed591e6ca9f4800/src/util/wpipe.cpp#L12-L38

Now wakeup pipes logic doesn't need to know what socket events mode is being used nor are the implementation aspects of (de)registering it its concern, that is now `EdgeTriggeredEvents` problem.

## Additional Information

* This pull request will need testing on macOS (FreeBSD isn't a tier-one target) to ensure that lack of breakage in `kqueue`-specific logic.

## Breaking Changes

* Dependency for https://github.com/dashpay/dash/pull/6018
* More logging has been introduced and existing log messages have been made more exhaustive. If there is parsing that relies on a particular template, they will have to be updated.
* If `EdgeTriggeredEvents` or `WakeupPipes` fail to initialize or are incorrectly initialized and not destroyed immediately, any further attempts at calling any of its functions will result in an `assert`-induced crash. Earlier behavior may have allowed for silent failure but segmentation of logic from `CConnman` means the newly created instances must only exist if the circumstances needed for it to initialize correctly are present.

  This is to ensure that `CConnman` doesn't have to concern itself with internal workings of either entities.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests **(note: N/A)**
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

**Comment by MrDefacto** [comment]
I can do FreeBSD testing. I know that there are a couple hundred masternodes running on FreeBSD. I don't think it's unimportant.

**Comment by UdjinM6** [comment]
works with no issues on mac (kqueue mode) it seems, a couple of suggestions: e0deb9e653 and maybe 4e7f80e1af

**Comment by kwvg** [comment]
@MrDefacto Messaged you on Keybase!

**Comment by kwvg** [comment]
> a couple of suggestions: https://github.com/dashpay/dash/commit/e0deb9e653cb01edcd9c5a4536a1a6f2bbfa1dad and maybe https://github.com/dashpay/dash/commit/4e7f80e1af5896bcb6ce60d423cadaeea77642f0

With regards to e0deb9e653cb01edcd9c5a4536a1a6f2bbfa1dad , I was originally planning to move `SocketEventsMode` to `util/sock.h` in [`net_processing_6`](https://github.com/kwvg/dash/tree/net_processing_6) (see https://github.com/kwvg/dash/commit/ed1409d88c65092b3d6a1f44d0cea30885abcd62) but getting rid of `EdgeEventsMode` and the need to sync enums was a good enough to reason to cherry-pick it. Thanks for the suggestion!

As for 4e7f80e1af5896bcb6ce60d423cadaeea77642f0, I'm leaning in the opposite direction and attempting to limit `CConnman`'s knowledge of the macro (this PR still has a few uses of `USE_`{`EPOLL`, `KQUEUE`, `POLL`} that is done away with in the successor PR). The latest push (d5f0838) removes a few more `USE_WAKEUP_PIPES` usage and eventually am hoping to contain it to `util/wpipes.cpp` entirely.

---

Also, a condition inversion in `WakeupPipe::Drain()` has been resolved in the latest push (for some reason the infinite loop introduced didn't manifest in this PR but did in [`net_processing_6`](https://github.com/kwvg/dash/tree/net_processing_6)).

**Review by UdjinM6 — APPROVED** [review]
LGTM, light ACK (on mac)

File: src/net.cpp [hunk 1]
```diff
@@ -3301,25 +3301,15 @@ bool CConnman::Start(CDeterministicMNManager& dmnman, CMasternodeMetaMan& mn_met
...
-        }
-    }
-#endif
-
-#ifdef USE_EPOLL
-    if (socketEventsMode == SocketEventsMode::EPoll) {
-        epollfd = epoll_create1(0);
-        if (epollfd == -1) {
-            LogPrintf("epoll_create1 failed\n");
+    std::optional<std::string> error;
```
Comment (PastaPastaPasta) [inline src/net.cpp:3273]: why was this added?

File: src/util/edge.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        if (m_fd == -1) {
+            LogPrintf("Unable to initialize EdgeTriggeredEvents, kqueue returned -1\n");
+            return;
+        }
+#else
+        LogPrintf("Attempting to initialize EdgeTriggeredEvents for kqueue without support compiled in!\n");
+        return;
+#endif /* USE_KQUEUE */
+    } else {
+        assert(false);
```
Comment (PastaPastaPasta) [inline src/util/edge.cpp:46]: why not use something like `__builtin_unreachable`?

File: src/util/edge.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    m_valid = true;
+}
+
+EdgeTriggeredEvents::~EdgeTriggeredEvents()
+{
+    if (m_valid) {
+#if defined(USE_KQUEUE) || defined(USE_EPOLL)
+        close(m_fd);
+#else
+        assert(false);
```
Comment (PastaPastaPasta) [inline src/util/edge.cpp:60]: why not use something like __builtin_unreachable?

File: src/util/sock.h [hunk 4]
```diff
@@ -18,6 +18,54 @@
...
+#endif /* USE_KQUEUE */
+        default:
+            return "unknown";
+    };
+}
+
+/* Converts string to SocketEventsMode value with additional check to report modes not compiled for as unknown */
+static SocketEventsMode SEMFromString(const std::string str)
+{
+         if (str == "select") { return SocketEventsMode::Select; }
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/util/sock.h]: Resolved in latest push

File: src/net.cpp (hunk 1 above)
Comment (kwvg) [reply src/net.cpp]: Leftover from early draft of PR, removed in latest push.

File: src/util/edge.cpp (hunk 2 above)
Comment (kwvg) [reply src/util/edge.cpp:46]: `__builtin_unreachable` is a compiler hint to optimize it out but is within the realm of possibility that the path might be taken due to erroneous code. As it is not an assertion that the code path isn't taken, `assert` seemed more suitable for the job.

**Review by PastaPastaPasta — APPROVED** [review]
utACK bd8b5d40073bcc54bd29663f06d965fafe349b08
